### PR TITLE
Prisma SQL

### PR DESCRIPTION
**Pull request by Itrulia**
Explanation of a few things:

1. I turned a lot of non nullable strings into nullable strings as they have been filled with an empty string.
2. On an update query in Prisma we can delete relations, but if that relation does not exist this throws an error. That's why I sometimes check with `Boolean(...)`
3. Nested objects such as properties or invoice items get deleted and recreated to keep compatibility. While they used to be nested, they are now their own table with ids. Current API does not expect to get an id for those, so we have no way to know which ones are which. So we just re-create them. This is ideally something we will change in the future. 
4. ArticleRevisions and PageRevisions are technically bad SQL as the Article has a revision instead of the Revision having an Article. This is kept for now to not make too many breaking changes
5. Tests have changed to generate unique inputs. This allows them to run without having to clear the database
6. Tests do not recreate the admin user anymore but seeds them now
7. MongoDB is still used by the oauth2 package

BREAKING CHANGE: @wepublish/api is now using Postgres instead of MongoDB

### CODE REVIEW

**Review by camilleory — COMMENTED** [review]
Good Job Karl 🥇 👍 !

File: config/create-gke-manifest.js [hunk 1]
```diff
@@ -56,6 +56,7 @@ async function main() {
   await applyEditor()
   await applyOAuth2()
   await applyMongo()
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply config/create-gke-manifest.js:58]: Yes for Oauth2

File: config/create-gke-manifest.js [hunk 2]
```diff
@@ -1327,6 +1329,147 @@ async function applyOAuth2() {
...
+              name: appName,
+              image: 'bitnami/postgresql:latest',
+              env: [
+                {
+                  name: "POSTGRESQL_DATABASE",
+                  value: "wepublish"
+                },
+                {
+                  name: "ALLOW_EMPTY_PASSWORD",
+                  value: "yes"
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply config/create-gke-manifest.js:1402]: Why "yes" or why empty password?
1st that's what Postgres uses for w/e reason, 2nd that's the same thing we do with Mongodb right now

File: docker-compose.yml [hunk 3]
```diff
@@ -54,14 +59,43 @@ services:
...
+      PGADMIN_CONFIG_MASTER_PASSWORD_REQUIRED: 'False'
+      PGADMIN_CONFIG_SERVER_MODE: 'False'
+    ports:
+      - "8000:80"
+    depends_on:
+      - database
+    networks:
+      - app-tier
+
   mongo:
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply docker-compose.yml:99]: For oauth2

File: docs/prisma.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+## Reset database
+
+You can reset the database by running the following command:
+
+```sh
+npx prisma migrate reset
+```
+
+This will completely wipe the database, migrate and seed it.
+
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply docs/prisma.md:185]: Thanks!

File: packages/api-media-karma/__tests__/specs/index.ts [hunk 5]
```diff
@@ -162,11 +172,12 @@ describe('Karma Media Adapter', () => {
     })
 
     test('GetImageURL should return a string with transformation', async () => {
-      const testImageWithFocal: Image = {
+      const testImageWithFocal: ImageWithFocalPoint = {
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply packages/api-media-karma/__tests__/specs/index.ts:175]: As SQL is mostly flat and FocalPoint was an object, it's now in a different table.
Image itself does not contain the focal point anymore, but it has to be added via `includes: {focalPoint: true}`, this will create a type `Image & {focalPoint: FocalPoint}` by Prisma, to reuse it and not having to set that everytime, I added `ImageWithFocalPoint`

File: packages/api/__tests__/specs/__snapshots__/page.ts.snap [hunk 6]
```diff
@@ -10,125 +10,8 @@ Object {
...
-        "publishedAt": null,
-        "slug": "testing-page",
-        "tags": Array [
-          "one",
-          "two",
-          "three",
-        ],
-        "title": "Testing Page",
-        "updatedAt": null,
+      "latest": ObjectContaining {
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply packages/api/__tests__/specs/__snapshots__/page.ts.snap:13]: I rewrote the tests so you can run them multiple times without having to clear the database, as clearing databases (especially just deleting and recreating) is very costly and sometimes clearing the database can fail but it's technically not test relevant.

So everything that was unique before, is now autogenerated, this is reflected in the snapshots.
I didn't check the snapshots yet to see if they are hiding relevant data.

File: docs/prisma.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+In that space you can add the path to the schema and seeding that should be executed.
+
+```json
+"prisma": {
+  "schema": "@wepublish/api/prisma/schema.prisma",
+  "seed": "npx ts-node seed.ts"
+},
+```
+
+To automate the generating of the primsa schema, we recommend to add the following script to your `package.json`:
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply docs/prisma.md:18]: Thank you!

File: packages/api/__tests__/api/private/author.graphql [hunk 8]
```diff
@@ -52,5 +52,7 @@ mutation UpdateAuthor($id: ID!, $input: AuthorInput!) {
 }
 
 mutation DeleteAuthor($id: ID!) {
-  deleteAuthor(id: $id)
+  deleteAuthor(id: $id) {
+    ...FullAuthor
```
Comment (penk85) [inline packages/api/__tests__/api/private/author.graphql:56]: Why did you add this? Consistency?

File: packages/api/__tests__/specs/article.ts [hunk 9]
```diff
@@ -87,41 +89,31 @@ describe('Articles', () => {
...
-              endCursor: expect.any(String),
-              startCursor: expect.any(String)
-            },
-            totalCount: expect.any(Number)
-          }
-        }
-      })
-      expect(articles.data?.articles?.totalCount).toBe(articleIds.length)
+
+      expect(res.data.articles.nodes).not.toHaveLength(0)
```
Comment (penk85) [inline packages/api/__tests__/specs/article.ts:116]: :+1:

File: packages/api/__tests__/specs/article.ts [hunk 10]
```diff
@@ -171,10 +166,14 @@ describe('Articles', () => {
...
         }
       })
+
       expect(updatedArticle).toMatchSnapshot({
         data: {
           updateArticle: {
-            id: expect.any(String)
+            id: expect.any(String),
+            latest: expect.objectContaining({
+              slug: expect.any(String)
```
Comment (penk85) [inline packages/api/__tests__/specs/article.ts:175]: You're simplifying it here because there's no need to test for it matching exact case?

File: packages/api/__tests__/api/private/author.graphql (hunk 8 above)
Comment (Itrulia) [reply packages/api/__tests__/api/private/author.graphql:56]: Deletion returns the old object now instead of `id` or `null`. Allows to undo deletions.

File: packages/api/__tests__/specs/article.ts (hunk 10 above)
Comment (Itrulia) [reply packages/api/__tests__/specs/article.ts:175]: See https://github.com/wepublish/wepublish/pull/644#discussion_r929926385

File: .env.example [hunk 11]
```diff
@@ -0,0 +1 @@
+DATABASE_URL=postgresql://postgres@localhost:5432/wepublish?schema=public
```
Comment (Itrulia) [inline .env.example:1]: Root level .env so that you can use prisma commands on root level

File: docker-compose.yml [hunk 12]
```diff
@@ -70,12 +104,16 @@ services:
     image: mongo-express
     restart: always
     ports:
-      - 8000:8081
+      - 8001:8081
```
Comment (Itrulia) [inline docker-compose.yml:107]: Keeping mongo-express for oauth2. Oauth2 will be deleted in the future as unused

File: examples/website/nodemon.json [hunk 13]
```diff
@@ -3,10 +3,10 @@
   "ignoreRoot": [".git"],
   "delay": 500,
   "env": {
-    "PORT": 5000,
```
Comment (Itrulia) [inline examples/website/nodemon.json:6]: Moved to port 4200 as port 5000 is used by the mac control center.

File: examples/website/package.json [hunk 14]
```diff
@@ -6,7 +6,7 @@
   "license": "UNLICENSED",
   "engines": {
     "node": "^12.0.0",
-    "yarn": "^1.17.0"
```
Comment (Itrulia) [inline examples/website/package.json:9]: Was relying on a different yarn version than the other packages

File: package.json [hunk 15]
```diff
@@ -7,34 +7,27 @@
...
     "prerelease": "run-s clean build:production",
     "release": "config/release.sh",
     "changelog": "node -r dotenv/config ./node_modules/.bin/lerna-changelog",
     "publish": "lerna publish from-package --ignore-scripts",
     "publish:next": "lerna publish from-package --ignore-scripts --dist-tag next",
     "watch": "run-p --print-label watch:*",
     "lint": "run-p --print-label lint:*",
-    "start:docker": "docker-compose up mongo mongo-express media",
+    "setup:root": "shx cp .env.example .env",
+    "start:docker": "docker compose up --detach mongo mongo-express database pgadmin media",
```
Comment (Itrulia) [inline package.json:25]: Starts detached now for better CLI output from our services.

File: package.json [hunk 16]
```diff
@@ -95,10 +88,12 @@
     "prettier": "^2.0.1",
     "pretty-quick": "^1.11.1",
     "prisma": "^3.14.0",
-    "testcafe": "^1.8.5"
+    "testcafe": "^1.8.5",
+    "ts-node": "^10.8.1"
```
Comment (Itrulia) [inline package.json:92]: Used for prisma seeding

File: packages/api-media-karma/__tests__/specs/index.ts [hunk 17]
```diff
@@ -31,7 +31,17 @@ const TEST_UPLOAD_IMAGE: Image = {
   mimeType: 'image/jpeg',
   format: 'testFormat',
   width: 1024,
-  height: 1024
+  height: 1024,
```
Comment (Itrulia) [inline packages/api-media-karma/__tests__/specs/index.ts:34]: Fix fixtures for types

File: packages/editor/src/client/atoms/invoice.tsx [hunk 18]
```diff
@@ -32,13 +32,14 @@ export function Invoice({subscriptionId, invoice, me, disabled, onInvoicePaid}:
...
     }
+
     // talk with the private api
     const items = prepareInvoiceItemsForApi(invoice.items)
     await updateInvoice({
       variables: {
         updateInvoiceId: invoice.id,
         input: {
-          items,
+          items: items.map(({total, ...item}) => item),
```
Comment (Itrulia) [inline packages/editor/src/client/atoms/invoice.tsx:42]: totals is calculated on graphql side and shouldn't be sent to the server

File: packages/api/prisma/migrations/20220713133350_init/migration.sql [hunk 19]
```diff
@@ -0,0 +1,745 @@
...
+CREATE TYPE "CommentRejectionReason" AS ENUM ('misconduct', 'spam');
+
+-- CreateEnum
+CREATE TYPE "CommentState" AS ENUM ('approved', 'pendingApproval', 'pendingUserChanges', 'rejected');
+
+-- CreateEnum
+CREATE TYPE "CommentAuthorType" AS ENUM ('team', 'author', 'verifiedUser', 'guestUser');
+
+-- CreateEnum
+CREATE TYPE "MailLogState" AS ENUM ('submitted', 'accepted', 'delivered', 'deferred', 'bounced', 'rejecte');
```
Comment (penk85) [inline packages/api/prisma/migrations/20220713133350_init/migration.sql:14]: "rejected" -typo

File: packages/api/src/graphql/peer/peer.private-mutation.ts [hunk 20]
```diff
@@ -40,7 +40,7 @@ export const updatePeer = (
   authorise(CanCreatePeer, roles)
 
   const nonEmptyInputs = Object.fromEntries(
-    Object.entries(input).filter(([, value]) => !value || value === false)
+    Object.entries(input).filter(([, value]) => value || value === false)
```
Comment (penk85) [inline packages/api/src/graphql/peer/peer.private-mutation.ts:43]: Is this intentional?

Comment (Itrulia) [reply packages/api/src/graphql/peer/peer.private-mutation.ts:43]: Yeah, the old logic was wrong (was a refactor from master to begin with, so this is not a bug on master)

File: packages/api/prisma/migrations/20220713133350_init/migration.sql [hunk 21]
```diff
@@ -0,0 +1,745 @@
...
+ALTER TABLE "payments" ADD CONSTRAINT "payments_paymentMethodID_fkey" FOREIGN KEY ("paymentMethodID") REFERENCES "payment.methods"("id") ON DELETE RESTRICT ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "peerProfiles" ADD CONSTRAINT "peerProfiles_logoID_fkey" FOREIGN KEY ("logoID") REFERENCES "images"("id") ON DELETE SET NULL ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "sessions" ADD CONSTRAINT "sessions_userID_fkey" FOREIGN KEY ("userID") REFERENCES "users"("id") ON DELETE RESTRICT ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "subscriptions.periods" ADD CONSTRAINT "subscriptions.periods_invoiceID_fkey" FOREIGN KEY ("invoiceID") REFERENCES "invoices"("id") ON DELETE CASCADE ON UPDATE CASCADE;
```
Comment (tomaszdurka) [inline packages/api/prisma/migrations/20220713133350_init/migration.sql:721]: Shouldn't we `set null` here? and allow null for period? or its always connected.

File: packages/api/prisma/migrations/20220713133350_init/migration.sql [hunk 22]
```diff
@@ -0,0 +1,745 @@
...
+ALTER TABLE "comments.revisions" ADD CONSTRAINT "comments.revisions_commentId_fkey" FOREIGN KEY ("commentId") REFERENCES "comments"("id") ON DELETE SET NULL ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "comments" ADD CONSTRAINT "comments_userID_fkey" FOREIGN KEY ("userID") REFERENCES "users"("id") ON DELETE SET NULL ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "invoices.items" ADD CONSTRAINT "invoices.items_invoiceId_fkey" FOREIGN KEY ("invoiceId") REFERENCES "invoices"("id") ON DELETE CASCADE ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "invoices" ADD CONSTRAINT "invoices_subscriptionID_fkey" FOREIGN KEY ("subscriptionID") REFERENCES "subscriptions"("id") ON DELETE CASCADE ON UPDATE CASCADE;
```
Comment (tomaszdurka) [inline packages/api/prisma/migrations/20220713133350_init/migration.sql:688]: Shouldn't we `set null` here? Leave invoice even if sub is deleted?

File: packages/api/prisma/schema.prisma [hunk 23]
```diff
@@ -32,81 +50,106 @@ type ArticleRevision {
...
-  @@map("articles")
-}
+  id         String   @id @default(cuid())
+  createdAt  DateTime @default(now())
+  modifiedAt DateTime @updatedAt
+
+  // The better way in the future to go around draft, pending, published is to have the articleId and a state on the revision
+  // There should then be a unique constraint on id + state.
+  // This would allow us to
+  //  cascade the revisions on article delete
```
Comment (tomaszdurka) [inline packages/api/prisma/schema.prisma:79]: Wdyt about keeping revisions forever ? And indeed deletion would just delete the wrapper object?

File: packages/api/prisma/schema.prisma [hunk 24]
```diff
@@ -117,185 +160,221 @@ model Image {
...
-  delivered  @map("delivered")
-  deferred   @map("deferred")
-  bounced    @map("bounced")
-  rejected   @map("rejected")
+  submitted @map("submitted")
+  accepted  @map("accepted")
+  delivered @map("delivered")
+  deferred  @map("deferred")
+  bounced   @map("bounced")
+  rejected  @map("rejecte")
```
Comment (tomaszdurka) [inline packages/api/prisma/schema.prisma:275]: `rejecte`?

File: packages/api/src/graphql/article.ts [hunk 25]
```diff
@@ -119,10 +119,10 @@ export const GraphQLArticleRevision = new GraphQLObjectType<ArticleRevision, Con
     hideAuthor: {type: GraphQLNonNull(GraphQLBoolean)},
 
     preTitle: {type: GraphQLString},
-    title: {type: GraphQLNonNull(GraphQLString)},
+    title: {type: GraphQLString},
```
Comment (tomaszdurka) [inline packages/api/src/graphql/article.ts:122]: Why nullable now/ missed that.

**Review by tomaszdurka — CHANGES_REQUESTED** [review]

File: packages/api/src/graphql/article.ts (hunk 25 above)
Comment (Itrulia) [reply packages/api/src/graphql/article.ts:122]: A lot of fields were filled with empty string by default. Made them nullable now.

File: packages/api/prisma/schema.prisma (hunk 23 above)
Comment (Itrulia) [reply packages/api/prisma/schema.prisma:79]: I totally agree with this, will need a few changes before we can do that (the way publishing, updating, etc. works)

File: docker-compose.yml [hunk 26]
```diff
@@ -54,14 +59,43 @@ services:
...
+
+  database:
+    image: bitnami/postgresql:latest
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply docker-compose.yml:72]: Forgot to do that, thanks will do!

File: docker-compose.yml [hunk 27]
```diff
@@ -1,13 +1,18 @@
...
   website:
     build: .
     ports:
-      - 5000:8000
+      - 4200:8000
     command: ['node', './examples/website/dist/server/index.js']
     environment:
       NODE_ENV: production
-      CANONICAL_HOST: http://localhost:5000
+      CANONICAL_HOST: http://localhost:4200
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply docker-compose.yml:15]: 5000 is used by the mac notification center, so you could never open it :D

**Review by penk85 — APPROVED** [review]
:clap:

File: .github/workflows/test.yml [hunk 28]
```diff
@@ -9,19 +9,22 @@ on:
 jobs:
   Test:
     runs-on: ubuntu-latest
+    services:
+      postgres:
+        image: bitnami/postgresql:latest
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply .github/workflows/test.yml:14]: We should lock them regardless, Prisma supports all of the current ones.
Will do! Thank you!

File: package.json (hunk 15 above)
Comment (Itrulia) [reply package.json:25]: Mongo is still needed for the oauth2 package

File: packages/api/__tests__/specs/comment.ts [hunk 29]
```diff
@@ -5,7 +5,7 @@
 
 beforeAll(async () => {
   try {
-    // const setupClient = await createGraphQLTestClientWithMongoDB()
+    // const setupClient = await createGraphQLTestClientWithPrisma()
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply packages/api/__tests__/specs/comment.ts:8]: Yeah whole file is really not needed. It ain't doing anything haha :D

File: docker-compose.yml (hunk 12 above)
Comment (michael-scheurer) [reply docker-compose.yml:107]: jira ticket created?

**Review by michael-scheurer — APPROVED** [review]
Found the following:
- yarn watch doesn't setup the database properly, which would be expected by docs. So either enhance docs or scripts.
- Creation of new Subscription from within editor not possible. Error: Cannot return null for non-nullable field UserSubscription.periods.
- Settings => Default values are not overtaken/seeded anymore.

You may fix it and then merge.

File: docker-compose.yml (hunk 27 above)
[the remark this answers is not in the document]
Comment (tomaszdurka) [reply docker-compose.yml:15]: Oh, never seen this conflict. At some point we set those port explicitly, but your change is fine for me.

[the remark this answers is not in the document]
Comment (Itrulia) [reply docker-compose.yml:15]: Was required to test the public api haha, that's why I kept it in there

**Review by tomaszdurka — APPROVED** [review]
